README.md
# OSM Map Widget for Elementor
A free Elementor Map Widget that utilizes [Open Street Map](https://www.openstreetmap.org/). Comes with features like adding multiple markers, and choosing from a library of custom tiles to change the look and feel. Requires [Elementor Page Builder](https://wordpress.org/plugins/elementor/) Plugin Version: 3.0.5+

This plugin is [open sourced](https://github.com/flopperj/osm-map-elementor) so feel free to contribute to it :)

## Usage
1. Make sure that **[Elementor Page Builder](https://wordpress.org/plugins/elementor/)** plugin is installed
2. Add optional API Keys to the settings page:
    - **Google Maps API Key**<br />_(Used to autocomplete and update the coordinates of markers in a map. Need help to get a Google map API key? [Read this resource](https://developers.google.com/maps/documentation/javascript/get-api-key))_
    - **Mapbox Access Token**<br />_(Used for custom tiles. Need help to get a Mapbox Access Token? [Read this resource](https://docs.mapbox.com/help/how-mapbox-works/access-tokens/))_
    - **Geoapify API Key**<br />_(Used for custom tiles. Need help to get a Geoapify API key? [Read this resource](https://www.geoapify.com/maps-api/))_ <br />
    <img src="https://raw.githubusercontent.com/flopperj/osm-map-elementor/master/assets/screenshot-1.jpg" width="500" height="auto" />
3. Add OSM Map to your elementor page and have fun :)
    - Find OSM Map widget from elements listing and add it to your elementor page<br />
      <img src="https://raw.githubusercontent.com/flopperj/osm-map-elementor/master/assets/screenshot-2.jpg" width="500" height="auto" />
    - Add multiple markers, change Zoom level and style them accordingly<br />
      <img src="https://raw.githubusercontent.com/flopperj/osm-map-elementor/master/assets/screenshot-3.jpg" width="500" height="auto" />
    - _Editing the markers:_ A marker's settings contains **Title, Location, Coordinates, Description, Button Text, and Button URL** fields. Of those fields, only the **Coordinates** field is required to render a marker on the map. If you have a Google Maps API key added, you'll be able to automatically populate the coordinates field without manually doing so.<br />
      <img src="https://raw.githubusercontent.com/flopperj/osm-map-elementor/master/assets/screenshot-4.jpg" width="250" height="auto" />
    - Change Tiles from style section _(requires **Mapbox Access Token** and **Geoapify API Key**)_<br />
      <img src="https://raw.githubusercontent.com/flopperj/osm-map-elementor/master/assets/screenshot-5.jpg" width="500" height="auto" />
## Licence
[GNU GENERAL PUBLIC LICENSE](https://raw.githubusercontent.com/flopperj/osm-map-elementor/master/LICENSE.txt)


CHANGELOG.md
# Changelog
Checkout what's new with OSM map widget for Elementor.

## Version 1.0.2
- Added more styling controls for markers.
- Fixed issue where button could not show up if description wasn't added

## Version 1.0.1
New:
 - Introduced OSM map widget for Elementor
 - Added ability to add multiple markers to the map.
 - Added ability to use custom map tiles through integration with Mapbox and Geoapify

elementor-osm-map.php
<?php
/**
 * Plugin Name:     OSM Map Widget for Elementor
 * Description:     An Elementor Widget that creates an OSM Map. Requires Elementor Plugin Version: 3.0.5+
 * Author:          ACT Innovate, James Arama, Alex Hooten
 * Author URI:      https://github.com/flopperj/elementor-osm-map
 * Version:         1.0.1
 */

namespace OSM_Map;

use Elementor\Plugin;

if (!defined('ABSPATH')) exit; // Exit if accessed directly

// The Widget_Base class is not available immediately after plugins are loaded, so
// we delay the class' use until Elementor widgets are registered
add_action('elementor/widgets/widgets_registered', function () {
    require_once('osm-map.php');

    $osm_map = new Widget_OSM_Map();

    // Let Elementor know about our widget
    Plugin::instance()
        ->widgets_manager
        ->register_widget_type($osm_map);
});


// Add global settings for our widget
add_action('admin_menu', function () {
    add_options_page('OSM Map Widget', 'OSM Map Widget', 'manage_options', 'osm-map-elementor', function () {

        $action = !empty($_REQUEST['action']) ? $_REQUEST['action'] : null;
        if (!empty($action) && $action == 'save_settings' && isset($_REQUEST['osm_widget'])) {
            $osm_settings = isset($_REQUEST['osm_widget']) ? $_REQUEST['osm_widget'] : [];
            update_option('osm_widget', $osm_settings);
            wp_redirect($_SERVER['HTTP_REFERER'] . '&action=settings_saved');
        }

        $osm_settings = get_option('osm_widget');

        ?>
        <style type="text/css">
            #osm-map-settings {
                margin: 5px 15px 2px;
            }

            #osm-map-settings div#alert-message {
                margin-left: 0;
            }

            #osm-map-settings h2 {
                color: #6d7882;
                font-weight: 400;
                font-size: 22px;
            }

            #osm-map-settings .card-header h3.card-title {
                color: #6d7882;
                padding: 5px 20px;
                font-weight: 500;
            }

            #osm-map-settings .card {
                padding: 0;
                border-color: #f1f1f1;;
            }

            #osm-map-settings .card-content {
                padding: 20px;
                background-color: #f7f7f7;
            }

            #osm-map-settings .card-header {
                border-bottom: 1px solid #f1f1f1;
            }

            #osm-map-settings input[type="text"] {
                width: 100%;
            }

            #osm-map-settings .osm-button {
                background: #e14d43;
                border-color: #e14d43;
                color: #fff;
                display: inline-block;
                text-decoration: none;
                font-size: 13px;
                line-height: 2.15384615;
                min-height: 30px;
                margin: 0;
                padding: 0 10px;
                cursor: pointer;
                border-width: 1px;
                border-style: solid;
                -webkit-appearance: none;
                border-radius: 3px;
                white-space: nowrap;
                box-sizing: border-box;
            }

            #osm-map-settings .form-group {
                margin-bottom: 25px;
            }
        </style>

        <div id="osm-map-settings">
            <h2>Elementor OSM Map Widget Settings</h2>
            <?php if (!empty($_REQUEST['action']) && $_REQUEST['action'] == 'settings_saved'): ?>
                <div style="background-color: rgb(255, 251, 204);" id="alert-message" class="updated"><p>
                        <strong><?php echo __('Settings saved') ?>.</strong></p></div>
            <?php endif; ?>
            <form action="<?php echo $_SERVER['REQUEST_URI'] . '&action=save_settings'; ?>" method="post">
                <div class="form-group">
                    <div class="card">
                        <div class="card-header">
                            <h3 class="card-title">Google Maps API Key</h3>
                        </div>
                        <div class="card-content">
                            <p><em><strong>Note:</strong> This setting is required if you wish to use Google Maps to
                                    lookup location coordinates.</em></p>
                            <input type="text" name="osm_widget[gmaps_key]"
                                   value="<?php echo !empty($osm_settings['gmaps_key']) ? $osm_settings['gmaps_key'] : null; ?>"/>
                        </div>
                    </div>
                </div>
                <div class="form-group">
                    <div class="card">
                        <div class="card-header">
                            <h3 class="card-title">Mapbox Access Token</h3>
                        </div>
                        <div class="card-content">
                            <p><em><strong>Note:</strong> This setting is required if you wish to use custom Mapbox / Geoapify tiles.</em></p>
                            <input type="text" name="osm_widget[mapbox_token]"
                                   value="<?php echo !empty($osm_settings['mapbox_token']) ? $osm_settings['mapbox_token'] : null; ?>"/>
                        </div>
                    </div>
                </div>
                <div class="form-group">
                    <div class="card">
                        <div class="card-header">
                            <h3 class="card-title">Geoapify API Key</h3>
                        </div>
                        <div class="card-content">
                            <p><em><strong>Note:</strong> This setting is required if you wish to use custom Geoapify tiles.</em></p>
                            <input type="text" name="osm_widget[geoapify_key]"
                                   value="<?php echo !empty($osm_settings['geoapify_key']) ? $osm_settings['geoapify_key'] : null; ?>"/>
                        </div>
                    </div>
                </div>
                <div class="form-group">
                    <button type="submit" class="osm-button">Save Changes</button>
                </div>
            </form>
        </div>

        <?php

    });
});

// Require Elementor plugin to be installed and activated
add_action('admin_init', function () {
    if (is_admin() && current_user_can('activate_plugins') && !is_plugin_active('elementor/elementor.php')) {
        add_action('admin_notices', function () {
            ?>
            <div class="error"><p>Sorry, but <strong>Elementor OSM Map</strong> Plugin requires that
                    <strong>Elementor</strong> plugin to be
                    installed and active.</p></div><?php
        });

        deactivate_plugins(plugin_basename(__FILE__));

        if (isset($_GET['activate'])) {
            unset($_GET['activate']);
        }
    }
});

assets/js/osm-map-controls.js
/**
 * Called as a callback to google maps API to allow for autocompletion of coordinates
 */
function initOSMEditorControls() {
    jQuery(window).on(
        'elementor:init',
        function () {

            /**
             * Initiates autocomplete
             * @param input
             * @param callback
             */
            const initiateAutoComplete = (input, callback) => {
                if (typeof google === 'undefined') return;

                const autocomplete = new google.maps.places.Autocomplete(input);
                google.maps.event.addListener(autocomplete, 'place_changed', function () {
                    const place = autocomplete.getPlace();
                    if (typeof callback === 'function') {
                        callback(place);
                    }
                });
            };


            elementor.hooks.addAction('panel/open_editor/widget/osm-map-elementor', function (panel, model, view) {

                // if value changes in input fields, trigger an input event that will allow for data to be saved
                jQuery('input[data-setting="marker_coords"],input[data-setting="marker_title"], input[data-setting="size"]').change(function () {
                    jQuery(this).trigger('input');
                });

                // auto-populate our coordinates
                const initMarkerFields = () => {
                    const markers = jQuery('.elementor-control-marker_list .elementor-repeater-fields:visible');

                    // loop through all visible marker fields
                    jQuery.each(markers, function () {
                        const location_input = jQuery(this).find('input[data-setting="marker_location"]');
                        const coords_input = jQuery(this).find('input[data-setting="marker_coords"]');
                        initiateAutoComplete(document.getElementById(location_input[0].id), function (place) {
                            coords_input.val(`${place.geometry.location.lat()},${place.geometry.location.lng()}`);
                            coords_input.trigger('input');
                            location_input.trigger('input');
                        });
                    });
                };

                initMarkerFields();

                // initialize markers on more additions
                jQuery('.elementor-repeater-add:visible').click(function () {
                    setTimeout(initMarkerFields, 1000);
                });
            });
        }
    );
}


osm-map-elementor.php
<?php
/**
 * Plugin Name:     OSM Map Widget for Elementor
 * Description:     A free Elementor Map Widget that Utilizes Open Street Map. Comes with features like adding multiple markers, and choosing from a library of custom tiles to change the look and feel.
 * Author:          ACT Innovate, James Arama, Alex Hooten
 * Author URI:      https://github.com/flopperj/elementor-osm-map
 * Version:         1.0.2
 */

namespace OSM_Map;

use Elementor\Plugin;

if (!defined('ABSPATH')) exit; // Exit if accessed directly

require_once ('constants.php');

// The Widget_Base class is not available immediately after plugins are loaded, so
// we delay the class' use until Elementor widgets are registered
add_action('elementor/widgets/widgets_registered', function () {
    require_once('osm-map.php');

    $osm_map = new Widget_OSM_Map();

    // Let Elementor know about our widget
    Plugin::instance()
        ->widgets_manager
        ->register_widget_type($osm_map);
});


// Add global settings for our widget
add_action('admin_menu', function () {
    add_options_page('OSM Map Widget', 'OSM Map Widget', 'manage_options', 'osm-map-elementor', function () {

        // queue admin styles
        wp_register_style('osm-map-admin', plugins_url('/osm-map-elementor/assets/css/admin.css'));
        wp_enqueue_style('osm-map-admin');

        $action = !empty($_REQUEST['action']) && is_string($_REQUEST['action']) ? sanitize_key($_REQUEST['action']) : null;

        // grab settings, sanitize, validate and save them
        if (!empty($action) && $action == 'save_settings' && isset($_REQUEST['osm_widget']) && is_array($_REQUEST['osm_widget'])) {

            $input = isset($_REQUEST['osm_widget']) ? $_REQUEST['osm_widget'] : [];

            // sanitize user input
            $osm_settings = [
                'gmaps_key' => !empty($input['gmaps_key']) ? sanitize_text_field($input['gmaps_key']) : null,
                'mapbox_token' => !empty($input['mapbox_token']) ? sanitize_text_field($input['mapbox_token']) : null,
                'geoapify_key' => !empty($input['geoapify_key']) ? sanitize_text_field($input['geoapify_key']) : null,
            ];

            // save the sanitized data
            update_option('osm_widget', $osm_settings);

            // redirect to form with confirmation alert message
            wp_redirect($_SERVER['HTTP_REFERER'] . '&action=settings_saved');
        }

        $osm_settings = get_option('osm_widget');

        ?>
        <div id="osm-map-settings">
            <h2>OSM Map Widget Settings</h2>
            <?php if (!empty($_REQUEST['action']) && sanitize_key($_REQUEST['action']) == 'settings_saved'): ?>
                <div style="background-color: rgb(255, 251, 204);" id="alert-message" class="updated"><p>
                        <strong><?php echo __('Settings saved') ?>.</strong></p></div>
            <?php endif; ?>
            <form action="<?php echo $_SERVER['REQUEST_URI'] . '&action=save_settings'; ?>" method="post">
                <div class="form-group">
                    <div class="card">
                        <div class="card-header">
                            <h3 class="card-title">Google Maps API Key</h3>
                        </div>
                        <div class="card-content">
                            <p><em><strong>Note:</strong> This setting is required if you wish to use Google Maps to
                                    lookup location coordinates. Need help to get a Google map API key? <a
                                            href="https://developers.google.com/maps/documentation/javascript/get-api-key"
                                            target="_blank">Read this resource</a>.</em></p>
                            <input type="text" name="osm_widget[gmaps_key]"
                                   value="<?php echo !empty($osm_settings['gmaps_key']) ? esc_textarea(__($osm_settings['gmaps_key'], OSM_MAP_SLUG)) : null; ?>"/>
                        </div>
                    </div>
                </div>
                <div class="form-group">
                    <div class="card">
                        <div class="card-header">
                            <h3 class="card-title">Mapbox Access Token</h3>
                        </div>
                        <div class="card-content">
                            <p><em><strong>Note:</strong> This setting is required if you wish to use custom Mapbox /
                                    Geoapify tiles. Need help to get a Mapbox Access Token? <a
                                            href="https://docs.mapbox.com/help/how-mapbox-works/access-tokens/"
                                            target="_blank">Read this resource</a></em></p>
                            <input type="text" name="osm_widget[mapbox_token]"
                                   value="<?php echo !empty($osm_settings['mapbox_token']) ? esc_textarea(__($osm_settings['mapbox_token'], OSM_MAP_SLUG)) : null; ?>"/>
                        </div>
                    </div>
                </div>
                <div class="form-group">
                    <div class="card">
                        <div class="card-header">
                            <h3 class="card-title">Geoapify API Key</h3>
                        </div>
                        <div class="card-content">
                            <p><em><strong>Note:</strong> This setting is required if you wish to use custom Geoapify
                                    tiles. Need help to get a Geoapify API key. <a
                                            href="https://www.geoapify.com/maps-api/">Read this resource.</a></em></p>
                            <input type="text" name="osm_widget[geoapify_key]"
                                   value="<?php echo !empty($osm_settings['geoapify_key']) ? esc_textarea(__($osm_settings['geoapify_key'], OSM_MAP_SLUG)) : null; ?>"/>
                        </div>
                    </div>
                </div>
                <div class="form-group">
                    <button type="submit" class="osm-button">Save Changes</button>
                </div>
            </form>
        </div>

        <?php

    });
});

// Require Elementor plugin to be installed and activated
add_action('admin_init', function () {
    if (is_admin() && current_user_can('activate_plugins') && !is_plugin_active('elementor/elementor.php')) {
        add_action('admin_notices', function () {
            ?>
            <div class="error"><p>Sorry, but <strong>Elementor OSM Map</strong> Plugin requires that
                    <strong>Elementor</strong> plugin to be
                    installed and active.</p></div><?php
        });

        deactivate_plugins(plugin_basename(__FILE__));

        if (isset($_GET['activate'])) {
            unset($_GET['activate']);
        }
    }
});